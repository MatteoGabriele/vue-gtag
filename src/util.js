export const noop = function() {};

export function loadScript(url) {
  return new Promise((resolve, reject) => {
    var head = document.head;
    const script = document.createElement("script");
    script.src = url;
    script.setAttribute("rel", "preconnect");
    script.setAttribute("async", true);
    script.charset = "utf-8";

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
}

export function warn(msg, err) {
  console.warn("[vue-gtag] " + msg);

  if (err && err.stack) {
    console.warn(err.stack);
  }
}

export function isFn(item) {
  return typeof item === "function";
}

export function isObject(item) {
  return item && typeof item === "object" && !Array.isArray(item);
}

export function mergeDeep(target, ...sources) {
  if (!sources.length) {
    return target;
  }

  const source = sources.shift();

  if (!isObject(target) || !isObject(source)) {
    return;
  }

  for (const key in source) {
    if (isObject(source[key])) {
      if (!target[key]) {
        Object.assign(target, { [key]: {} });
      }

      mergeDeep(target[key], source[key]);
    } else {
      Object.assign(target, { [key]: source[key] });
    }
  }

  return mergeDeep(target, ...sources);
}
