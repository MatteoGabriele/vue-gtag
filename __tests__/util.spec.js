import * as util from "@/util";
import flushPromises from "flush-promises";

describe("loadScript", () => {
  afterEach(() => {
    document.getElementsByTagName("html")[0].innerHTML = "";
  });

  it("should return a promise", () => {
    expect(util.loadScript("a")).toBeInstanceOf(Promise);
  });

  it("should create a script tag", (done) => {
    util.loadScript("foo");

    flushPromises().then(() => {
      expect(document.head).toMatchSnapshot();
      done();
    });
  });

  it("should create a link for domain preconnect", (done) => {
    util.loadScript("foo", "bar");

    flushPromises().then(() => {
      expect(document.head).toMatchSnapshot();
      done();
    });
  });
});

describe("mergeDeep", () => {
  it("should merge two objects", () => {
    const a = { a: 1, c: { a: 1 } };
    const b = { b: 1, c: { b: 1 } };
    const c = util.mergeDeep(a, b);

    expect(c).toMatchObject({
      a: 1,
      b: 1,
      c: {
        b: 1,
        a: 1,
      },
    });
  });

  it("should merge objects containing arrays", () => {
    const a = { c: [1, 2, 3] };
    const b = { c: [4, 5], d: { b: 1 } };
    const c = util.mergeDeep(a, b);

    expect(c).toEqual({
      c: [4, 5],
      d: { b: 1 },
    });
  });
});

describe("warn", () => {
  it("should warn with a customized message prefix", () => {
    console.warn = jest.fn();
    util.warn("foo");
    expect(console.warn).toHaveBeenCalledWith("[vue-gtag] foo");
  });

  it("should console the error", () => {
    console.warn = jest.fn();
    util.warn("foo", new Error());
    expect(console.warn).toHaveBeenCalledTimes(2);
  });
});

describe("isFn", () => {
  it("should return true", () => {
    expect(util.isFn(() => {})).toBe(true);
  });

  it("should return false", () => {
    expect(util.isFn("hello")).toBe(false);
  });
});

describe("isObject", () => {
  it("should return true", () => {
    expect(util.isObject({})).toBe(true);
  });
  it("should return false", () => {
    expect(util.isObject([])).toBe(false);
    expect(util.isObject("aa")).toBe(false);
  });
});
