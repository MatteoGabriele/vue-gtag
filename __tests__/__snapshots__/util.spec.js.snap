// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadScript should create a link for domain preconnect 1`] = `
<head>
  <link
    href="bar"
    rel="preconnect"
  />
  <script
    charset="utf-8"
    src="foo"
  />
</head>
`;

exports[`loadScript should create a script tag 1`] = `
<head>
  <script
    charset="utf-8"
    src="foo"
  />
</head>
`;

exports[`loadScript should create a script tag with the defer attribute 1`] = `
<head>
  <script
    charset="utf-8"
    defer=""
    src="foo"
  />
</head>
`;
